File: ios/Utils/Customization.swift
//
//  Customization.swift
//  FaceTec
//
//  Created by Alex Serdukov on 22.12.2020.
//  Copyright © 2020 Facebook. All rights reserved.
//

final class Customization {
    public static let resultSuccessMessage = "You’re a beautiful\n& unique unicorn!"

    private static let isLargeDevice = UIScreen.main.bounds.width >= 395

    private static let black = UIColor.black
    private static let white = UIColor.white
    private static let green = UIColor.green
    private static let darkGray = UIColor.darkGray
    private static let lightGray = UIColor.lightGray

    private static let primary = UIColor("#00afff")
    private static let gray50Percent = UIColor("#cbcbcb")

    private static let defaultCornerRadius: Int32 = 5
    private static let defaultFont = UIFont.robotoFont(12)

    private static let defaultBundle: Bundle = {
        let podBundle = Bundle(for: Customization.self)
        let bundleUrl = podBundle.url(forResource: "FaceTec", withExtension: "bundle")!

        return Bundle(url: bundleUrl)!
    }()
  
    static let UICustomization = createBasicCustomization()

    static let LowLightModeCustomization: FaceTecCustomization = {
        let ui = createBasicCustomization()
        let guidance = ui.guidanceCustomization
        
        guidance.readyScreenHeaderTextColor = white
        guidance.readyScreenSubtextTextColor = white

        return ui
    }()
    
    static let DynamicModeCustomization: FaceTecCustomization = {
      let ui = createBasicCustomization()
      let guidance = ui.guidanceCustomization
      let cancelButton = ui.cancelButtonCustomization
      
      cancelButton.customImage = UIImage(named: "CancelWhite.png", in: defaultBundle, compatibleWith: nil)!
      guidance.readyScreenHeaderTextColor = black
      guidance.readyScreenSubtextTextColor = black

      return ui
    }()

    private static func createBasicCustomization() -> FaceTecCustomization {
      let ui = FaceTecCustomization()

      let oval = ui.ovalCustomization
      let frame = ui.frameCustomization
      let overlay = ui.overlayCustomization
      let feedback = ui.feedbackCustomization
      let guidance = ui.guidanceCustomization
      let resultScreen = ui.resultScreenCustomization
      let cancelButton = ui.cancelButtonCustomization

      // removing branding image from overlay
      overlay.showBrandingImage = false
      overlay.backgroundColor = white.withAlphaComponent(0.5)

      // setting custom location & image of cancel button
      cancelButton.location = .topRight
      cancelButton.customImage = UIImage(named: "Cancel.png", in: defaultBundle, compatibleWith: nil)!

      // configuring feedback bar typography & border radius
      feedback.backgroundColor = CAGradientLayer.solidFill(color: black)
      feedback.cornerRadius = defaultCornerRadius
      feedback.textColor = white
      feedback.textFont = UIFont.robotoFont(.bold, 24)

      // setting oval border color & width
      oval.strokeWidth = 6
      oval.strokeColor = primary
      oval.progressColor1 = green
      oval.progressColor2 = green

      // frame (zoom's popup) customizations
      // setting frame border, radius & shadow
      frame.borderColor = white.withAlphaComponent(0)
      frame.cornerRadius = defaultCornerRadius
      frame.borderWidth = 0
      frame.shadow = FaceTecShadow.css(boxShadow: [0, 19, 38, 0], black, 0.42)

      // setting Zoom UI background color
      frame.backgroundColor = white

      // guidance screens ("frame your face", "retry" etc) customizations
      // setting setting Zoom UI default text color
      guidance.foregroundColor = darkGray

      // customizing buttons
      guidance.buttonFont = defaultFont
      guidance.buttonBorderWidth = 0
      guidance.buttonCornerRadius = defaultCornerRadius
      guidance.buttonTextNormalColor = white
      guidance.buttonTextHighlightColor = white
      guidance.buttonTextDisabledColor = white
      guidance.buttonBackgroundNormalColor = primary
      guidance.buttonBackgroundHighlightColor = primary
      guidance.buttonBackgroundDisabledColor = primary

      // customizing header / subtext
      guidance.headerFont = UIFont.robotoFont(.medium, isLargeDevice ? 22 : 20)

      // subtext
      guidance.subtextFont = defaultFont

      // configuring guidance images on retry screen
      guidance.retryScreenOvalStrokeColor = primary
      guidance.retryScreenImageBorderColor = primary
      guidance.retryScreenImageBorderWidth = 4
      guidance.retryScreenImageCornerRadius = defaultCornerRadius

      // customizing result screen - progress bar & success animation
      resultScreen.foregroundColor = darkGray
      resultScreen.messageFont = defaultFont.withSize(16)
      resultScreen.showUploadProgressBar = true
      resultScreen.uploadProgressFillColor = primary
      resultScreen.uploadProgressTrackColor = lightGray
      resultScreen.resultAnimationBackgroundColor = white
      resultScreen.resultAnimationForegroundColor = primary
      resultScreen.customActivityIndicatorImage = UIImage(named: "ActivityIndicator.png", in: defaultBundle, compatibleWith: nil)!
      resultScreen.customActivityIndicatorRotationInterval = 3000

      return ui
    }
}

File: android/src/main/java/org/gooddollar/facetec/util/Customization.java
package org.gooddollar.facetec.util;

import android.graphics.Color;
import android.graphics.Typeface;
import androidx.annotation.Nullable;

import com.facetec.sdk.R;
import com.facetec.sdk.FaceTecCustomization;
import com.facetec.sdk.FaceTecOverlayCustomization;
import com.facetec.sdk.FaceTecCancelButtonCustomization;
import com.facetec.sdk.FaceTecFeedbackCustomization;
import com.facetec.sdk.FaceTecOvalCustomization;
import com.facetec.sdk.FaceTecFrameCustomization;
import com.facetec.sdk.FaceTecGuidanceCustomization;
import com.facetec.sdk.FaceTecResultScreenCustomization;
import com.facetec.sdk.FaceTecVocalGuidanceCustomization;

import java.util.Map;
import java.util.HashMap;

import org.gooddollar.facetec.R.drawable;

public class Customization {
  private Customization() {}

  final public static String resultSuccessMessage = "You’re a beautiful\n& unique unicorn!";

  final public static FaceTecCustomization UICustomization = basicCustomizationFactory();
  final public static FaceTecCustomization LowLightModeCustomization = basicCustomizationFactory();
  final public static FaceTecCustomization DynamicModeCustomization = basicCustomizationFactory();
  final public static Map<Integer, String> UITextStrings = new HashMap<>();

  final private static int black = Color.BLACK;
  final private static int white = Color.WHITE;
  final private static int whiteTransparent = 0x00ffffff;
  final private static int whiteSemiTransparent = 0x80ffffff;
  final private static int green = Color.GREEN;
  final private static int darkGray = Color.DKGRAY;
  final private static int lightGray = Color.LTGRAY;

  final private static int primary = 0xff00afff;
  final private static int gray50Percent = 0xffcbcbcb;

  final private static int defaultCornerRadius = 5;
  final private static int buttonCornerRadius = 12;

  // "sans-serif" family resolves to the Roboto font
  final private static Typeface defaultFont = Typeface.create("sans-serif", Typeface.NORMAL);
  final private static Typeface mediumFont = Typeface.create("sans-serif-medium", Typeface.NORMAL);
  final private static Typeface boldFont = Typeface.create("sans-serif", Typeface.BOLD);

  static {
    // customize Dynamic dimming & low light mode UI
    FaceTecGuidanceCustomization lowLightGuidance = LowLightModeCustomization.getGuidanceCustomization();
    FaceTecGuidanceCustomization dynamicDimmingGuidance = DynamicModeCustomization.getGuidanceCustomization();
    FaceTecCancelButtonCustomization dynamicDimmingButton = DynamicModeCustomization.getCancelButtonCustomization();

    // Dynamic dimming customizations
    dynamicDimmingButton.customImage = drawable.facetec_cancel_white;
    dynamicDimmingGuidance.readyScreenHeaderTextColor = black;
    dynamicDimmingGuidance.readyScreenSubtextTextColor = black;

    // Low light customizations
    lowLightGuidance.readyScreenHeaderTextColor = white;
    lowLightGuidance.readyScreenSubtextTextColor = white;
  }

  final private static FaceTecCustomization basicCustomizationFactory() {
    FaceTecCustomization customization = new FaceTecCustomization();

    // customize UI
    FaceTecOverlayCustomization overlay = customization.getOverlayCustomization();
    FaceTecCancelButtonCustomization cancelButton = customization.getCancelButtonCustomization();
    FaceTecFeedbackCustomization feedback = customization.getFeedbackCustomization();
    FaceTecOvalCustomization oval = customization.getOvalCustomization();
    FaceTecFrameCustomization frame = customization.getFrameCustomization();
    FaceTecGuidanceCustomization guidance = customization.getGuidanceCustomization();
    FaceTecVocalGuidanceCustomization vocalGuidance = customization.vocalGuidanceCustomization;
    FaceTecResultScreenCustomization resultScreen = customization.getResultScreenCustomization();

    // removing branding image from overlay
    overlay.showBrandingImage = false;
    overlay.backgroundColor = whiteSemiTransparent;

    // setting custom location & image of cancel button
    cancelButton.setLocation(FaceTecCancelButtonCustomization.ButtonLocation.TOP_RIGHT);
    cancelButton.customImage = drawable.facetec_cancel;

    // configuring feedback bar typography & border radius
    feedback.backgroundColors = primary;
    feedback.cornerRadius = defaultCornerRadius;
    feedback.textColor = white;
    feedback.textFont = boldFont;
    feedback.elevation = 5;

    // setting oval border color & width
    oval.strokeWidth = 6;
    oval.strokeColor = primary;
    oval.progressColor1 = green;
    oval.progressColor2 = green;

    // frame (zoom's popup) customizations
    // setting frame border, radius & elevation
    frame.borderColor = whiteTransparent;
    frame.cornerRadius = defaultCornerRadius;
    frame.borderWidth = 0;
    frame.elevation = 19;

    // setting Zoom UI background color
    frame.backgroundColor = white;

    // guidance screens ("frame your face", "retry" etc) customizations
    // setting setting Zoom UI default text color
    guidance.foregroundColor = darkGray;

    // customizing buttons
    guidance.buttonFont = defaultFont;
    guidance.buttonBorderWidth = 0;
    guidance.buttonCornerRadius = buttonCornerRadius;
    guidance.buttonTextNormalColor = white;
    guidance.buttonTextHighlightColor = white;
    guidance.buttonTextDisabledColor = white;
    guidance.buttonBackgroundNormalColor = primary;
    guidance.buttonBackgroundHighlightColor = primary;
    guidance.buttonBackgroundDisabledColor = primary;

    // customizing header / subtext
    guidance.headerFont = mediumFont;

    // subtext
    guidance.subtextFont = defaultFont;

    // configuring guidance images on retry screen
    guidance.retryScreenOvalStrokeColor = primary;
    guidance.retryScreenImageBorderColor = primary;
    guidance.retryScreenImageBorderWidth = 4;
    guidance.retryScreenImageCornerRadius = defaultCornerRadius;

    // customizing result screen - progress bar & success animation
    resultScreen.foregroundColor = darkGray;
    resultScreen.messageFont = defaultFont;
    resultScreen.showUploadProgressBar = true;
    resultScreen.uploadProgressFillColor = primary;
    resultScreen.uploadProgressTrackColor = lightGray;
    resultScreen.resultAnimationBackgroundColor = white;
    resultScreen.resultAnimationForegroundColor = primary;
    resultScreen.customActivityIndicatorImage = drawable.facetec_activity_indicator;
    resultScreen.customActivityIndicatorRotationInterval = 3000;

    // disable voice help
    vocalGuidance.mode = FaceTecVocalGuidanceCustomization.VocalGuidanceMode.NO_VOCAL_GUIDANCE;

    return customization;
  }
}
